### PR TITLE
Add limb-darkening generation with exotic-ld

### PR DESCRIPTION
**Pull request by evamariaa**
Generates limb-darkening coefficients (linear, quadratic, 3-parameter, 4-parameter) at Stage 4 using the exotic-ld package (needs senstivity files from exotic-ld zenodo link!) and saves it in the astreaus lc object.
At Stage 5 there's a parameter to choose to use it and then the fixed limb-darkening parameters from the .epf file are overwritten with the generated one. Also uses the limb_dark for defining which coefficients to use.

### CODE REVIEW

**Comment by taylorbell57** [comment]
It looks like the tests currently fail - perhaps there's something missing from the testing ECFs that you added to the other ECFs

File: src/eureka/S5_lightcurve_fitting/models/BatmanModels.py [hunk 1]
```diff
@@ -45,9 +45,24 @@ def __init__(self, **kwargs):
...
-        ld_func = ld_profile(self.parameters.limb_dark.value)
+        self.ld_from_S4 = kwargs.get('ld_from_S4')
+        ld_func = ld_profile(self.parameters.limb_dark.value, 
+                             use_gen_ld=self.ld_from_S4)
         len_params = len(inspect.signature(ld_func).parameters)
         self.coeffs = ['u{}'.format(n) for n in range(len_params)[1:]]
+        
+        # Replace fixed u parameters with generated limb-darkening values
+        if self.ld_from_S4 is not None:
+            self.ld_S4_array = kwargs.get('ld_coeffs')[len_params-2]
```
Comment (taylorbell57) [inline src/eureka/S5_lightcurve_fitting/models/BatmanModels.py:56]: I'm not sure what went wrong with the automated tests, but when I run the tests locally I get the following error at this line for the MIRI and NIRCam tests (didn't try the other instruments)
```
        # Replace fixed u parameters with generated limb-darkening values
        if self.ld_from_S4 is not None:
>           self.ld_S4_array = kwargs.get('ld_coeffs')[len_params-2]
E           TypeError: 'NoneType' object is not subscriptable

/Users/tjbell1/Research/JWST/ERS/Eureka/src/eureka/S5_lightcurve_fitting/models/BatmanModels.py:56: TypeError
```

File: src/eureka/S5_lightcurve_fitting/s5_fit.py [hunk 2]
```diff
@@ -181,21 +196,18 @@ def fitlc(eventlabel, ecf_path=None, s4_meta=None):
...
-                    
+                    flux = lc.data.values[channel, :]
+                    flux_err = lc.err.values[channel, :]
+
                     # Normalize flux and uncertainties to avoid large
                     # flux values
-                    flux_err = flux_err/np.ma.mean(flux)
-                    flux = flux/np.ma.mean(flux)
+                    flux_err = flux_err/np.nanmean(flux)
+                    flux = flux/np.nanmean(flux)
```
Comment (taylorbell57) [inline src/eureka/S5_lightcurve_fitting/s5_fit.py:205]: These lines should also be reverted to what is in the main branch

File: src/eureka/S5_lightcurve_fitting/s5_fit.py [hunk 3]
```diff
@@ -254,7 +267,7 @@ def fit_channel(meta, time, flux, chan, flux_err, eventlabel, sharedp, params,
     lc_model = lc.LightCurve(time, flux, chan, chanrng, log, longparamlist,
                              unc=flux_err, time_units=time_units,
                              name=eventlabel, share=sharedp)
-
+
```
Comment (taylorbell57) [inline src/eureka/S5_lightcurve_fitting/s5_fit.py:270]: There should be no spaces on this blank line for flake8

File: tests/MIRI_ecfs/S4_MIRI.ecf [hunk 4]
```diff
@@ -25,6 +25,15 @@ maxiters		5		# The number of iterations of sigma clipping that should be perform
 boundary        'fill'  # Use 'fill' to extend the boundary values by the median of all data points (recommended), 'wrap' to use a periodic boundary, or 'extend' to use the first/last data points
 fill_value		mask	# Either the string 'mask' to mask the outlier values (recommended), 'boxcar' to replace data with the mean from the box-car filter, or a constant float-type fill value.
 
+#Limb-darkening parameters needed to compute exotic-ld
+compute_ld      False
+inst_filter     LRS #filter of JWST/HST instrument, supported list see https://exotic-ld.readthedocs.io/en/latest/views/supported_instruments.html 
+metallicity     0.1 #metallicity of the star 
+teff            6000 #effective temperature of the star in K
+logg            4.0 #surface gravity in log g
+exotic_ld_direc /exotic-ld_data/ #directory for ancillary files for exotic-ld
```
Comment (taylorbell57) [inline tests/MIRI_ecfs/S4_MIRI.ecf:34]: Is this folder with respect to some other folder? Surely the files aren't being saved in the root directory of your computer?

File: tests/NIRSpec_ecfs/S4_NIRSpec.ecf [hunk 5]
```diff
@@ -24,6 +24,15 @@ box_width		10		# The width of the box-car filter (used to calculated the rolling
 maxiters		5		# The number of iterations of sigma clipping that should be performed.
 fill_value		mask	# Either the string 'mask' to mask the outlier values (recommended), 'boxcar' to replace data with the mean from the box-car filter, or a constant float-type fill value.
 
+#Limb-darkening parameters needed to compute exotic-ld
+compute_ld      False
+inst_filter     G395H #filter of JWST/HST instrument, supported list see https://exotic-ld.readthedocs.io/en/latest/views/supported_instruments.html
```
Comment (taylorbell57) [inline tests/NIRSpec_ecfs/S4_NIRSpec.ecf:29]: This should be PRISM

File: demos/JWST/S4_template.ecf [hunk 6]
```diff
@@ -27,6 +27,15 @@ maxiters        5       # The number of iterations of sigma clipping that should
 boundary        'fill'  # Use 'fill' to extend the boundary values by the median of all data points (recommended), 'wrap' to use a periodic boundary, or 'extend' to use the first/last data points
 fill_value      mask    # Either the string 'mask' to mask the outlier values (recommended), 'boxcar' to replace data with the mean from the box-car filter, or a constant float-type fill value.
 
+#Limb-darkening parameters needed to compute exotic-ld
+compute_ld      True
+inst_filter     F322W2 #filter of JWST/HST instrument, supported list see https://exotic-ld.readthedocs.io/en/latest/views/supported_instruments.html
```
Comment (taylorbell57) [inline demos/JWST/S4_template.ecf:32]: This should be PRISM since that's what the demos are supposed to be for

File: demos/JWST/S5_template.ecf [hunk 7]
```diff
@@ -11,8 +11,11 @@ verbose         True                    # If True, more details will be printed
 fit_method      [dynesty]               #options are: lsq, emcee, dynesty (can list multiple types separated by commas)
 run_myfuncs     [batman_tr,polynomial]  #options are: batman_tr, batman_ecl, sinusoid_pc, expramp, GP, and polynomial (can list multiple types separated by commas)
 
-# Limb darkening controls (not yet implemented)
-#fix_ld          False #use limb darkening file?
+# Limb darkening controls
+use_generate_ld  exotic-ld #use the generated limb-darkening coefficients from Stage 4? Options: exotic-ld
```
Comment (taylorbell57) [inline demos/JWST/S5_template.ecf:15]: Note that None is also an option here

**Review by taylorbell57 — CHANGES_REQUESTED** [review]
Alright, here are the edits I see so far (haven't yet looked at src/eureka/S4_generate_lightcurves/generate_LD.py). Also, the NIRCam test is currently failing because the sigma is ending up with no spread in dynesty - not sure what caused that

**Comment by taylorbell57** [comment]
The new ECF settings also need to be documented in docs/source/ecf.rst and the ECFs in demos/media also need to be updated

File: src/eureka/S4_generate_lightcurves/generate_LD.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+    # Compute wavelength range, needs to be in Angstrom
+    wsdata = np.array(meta.wave_low)
+    wsdata = np.append(wsdata, meta.wave_hi[-1])
+    wavelength_range = []
+    for i in range(meta.nspecchan):
+        wavelength_range.append([wsdata[i], wsdata[i+1]])
+    wavelength_range = np.array(wavelength_range)
+    if spec.wave_1d.attrs['wave_units'] == 'microns':
+        wavelength_range *= 1e4
```
Comment (taylorbell57) [inline src/eureka/S4_generate_lightcurves/generate_LD.py:44]: Should be dividing, not multiplying

Comment (taylorbell57) [reply src/eureka/S4_generate_lightcurves/generate_LD.py]: Hah, oops - not it should not

**Review by taylorbell57 — APPROVED** [review]
